sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java

```diff
@@ -0,0 +1,198 @@
+  }
+
+  @Override
+  public synchronized @Nullable ProfilingTraceData onTransactionFinish(
+      @NotNull ITransaction transaction) {
+
+    final ITransaction finalActiveTransaction = activeTransaction;
+    final ProfilingTraceData profilingData = timedOutProfilingData;
+
+    // Profiling finished, but we check if we cached last profiling data due to a timeout
+    if (finalActiveTransaction == null) {
+      // If the cached timed out profiling data refers to the transaction that started it we return
+      // it back, otherwise we would simply lose it
+      if (profilingData != null) {
+        // The timed out transaction is finishing
+        if (profilingData
+            .getTraceId()
+            .equals(transaction.getSpanContext().getTraceId().toString())) {
+          timedOutProfilingData = null;
+          return profilingData;
+        } else {
+          // Another transaction is finishing before the timed out one
+          options
+              .getLogger()
+              .log(
+                  SentryLevel.ERROR,
+                  "Profiling data with id %s exists but doesn't match the closing transaction %s",
+                  profilingData.getTraceId(),
+                  transaction.getSpanContext().getTraceId().toString());
+          return null;
+        }
+      }
+      // A transaction is finishing, but profiling didn't start. Maybe it was started by another one
+      options
+          .getLogger()
+          .log(
+              SentryLevel.INFO,
+              "Transaction %s finished, but profiling never started for it. Skipping",
+              transaction.getSpanContext().getTraceId().toString());
+      return null;
+    }
+
+    if (finalActiveTransaction != transaction) {
+      options
+          .getLogger()
+          .log(
+              SentryLevel.DEBUG,
+              "Transaction %s finished, but profiling was started by transaction %s. Skipping",
+              transaction.getSpanContext().getTraceId().toString(),
+              finalActiveTransaction.getSpanContext().getTraceId().toString());
+      return null;
+    }
+
+    Debug.stopMethodTracing();
+
+    activeTransaction = null;
+
+    if (scheduledFinish != null) {
+      scheduledFinish.cancel(true);
+      scheduledFinish = null;
+    }
+
+    if (traceFile == null || !traceFile.exists()) {
+      options
+          .getLogger()
+          .log(
+              SentryLevel.ERROR,
+              "Trace file %s does not exists",
+              traceFile == null ? "null" : traceFile.getPath());
+      return null;
+    }
+
+    return new ProfilingTraceData(
+        traceFile,
+        transaction,
+        Build.VERSION.SDK_INT,
+        Build.MANUFACTURER,
+        Build.MODEL,
+        Build.VERSION.RELEASE,
+        options.getProguardUuid(),
+        packageInfo != null ? ContextUtils.getVersionName(packageInfo) : "",
+        packageInfo != null ? ContextUtils.getVersionCode(packageInfo) : "",
```

Comment:
We could do a single check for `packageInfo` outside of this block, so we spare one comparison.

Reply:
I could do a single check for `PackageInfo`, but then i would need two boolean checks for version name and code.
I think a null check is fast enough not to cache this check, but i can do it if you think it's better to do so

Reply:
You'd rather assign to 2 different Strings, they are already with the value or "", this is just a nit.

```
String version = "";
String code "";
if (packageInfo != null) {
  // assign values to both strings
}
```